util/src/main/java/com/google/cloud/healthcare/imaging/dicomadapter/AttributesUtil.java

```diff
@@ -0,0 +1,219 @@
+      // non-string type search keys don't seem to exist
+      // pass-through multiple values (valid for UID lists, ignored for rest)
+      for (String value : attrs.getStrings(keyTag)) {
+        String encodedValue;
+        try {
+          encodedValue = URLEncoder.encode(value, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+          throw new DicomServiceException(Status.ProcessingFailure, e);
+        }
+        qidoPath.append(TagUtils.toHexString(keyTag) + "=" + encodedValue + "&");
+      }
+    }
+
+    return qidoPath.toString();
+  }
+
+  /**
+   * Converts QIDO_RS json reply to dcm4che Attributes
+   * @param jsonDataset
+   */
+  public static Attributes jsonToAttributes(JSONObject jsonDataset) throws DicomServiceException {
+    Attributes attributes = new Attributes();
+    for (String tag : jsonDataset.keySet()) {
+      JSONObject element = jsonDataset.getJSONObject(tag);
+      int tagInt = TagUtils.forName(tag);
+      VR vr = VR.valueOf(element.getString("vr"));
+      if (element.has("Value")) {
+        setAttributeValue(attributes, tagInt, vr, (JSONArray) element.get("Value"));
+      } else if (element.has("DataFragment") && isBinaryVr(vr)) {
+        setAttributeDataFragment(attributes, tagInt, vr, element);
+      } else {
+        attributes.setValue(tagInt, vr, null);
+      }
+    }
+    return attributes;
+  }
+
+  private static boolean isBinaryVr(VR vr) {
+    switch (vr) {
+      case OB:
+      case OD:
+      case OF:
+      case OL:
+      case OW:
+      case UN:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private static void setAttributeValue(Attributes attrs, int tag, VR vr, JSONArray jsonValues)
+      throws DicomServiceException {
+    if (vr.isStringType() || vr == VR.AT) {
+      String[] values = jsonValues.toList().stream().map(Object::toString)
+          .collect(Collectors.toList()).toArray(new String[]{});
+      attrs.setString(tag, vr, values);
+    } else if (vr.isIntType()) {
+      int[] values = jsonValues.toList().stream().mapToInt(i -> Integer.parseInt(i.toString()))
+          .toArray();
+      attrs.setInt(tag, vr, values);
+    } else {
+      switch (vr) {
+        case SQ:
+          Sequence sequence = attrs.newSequence(tag, jsonValues.length());
+          for (Object seqElement : jsonValues) {
+            Attributes seqElementAttributes = jsonToAttributes((JSONObject) seqElement);
+            sequence.add(seqElementAttributes);
+          }
+          break;
+        case FL:
+        case FD:
+          double[] dValues = jsonValues.toList().stream().mapToDouble(i -> (Double) i).toArray();
+          attrs.setDouble(tag, vr, dValues);
+          break;
+        default:
+          throw new DicomServiceException(Status.ProcessingFailure,
+              "Unsupported VR " + vr.toString());
+      }
+    }
+  }
+
+  private static void setAttributeDataFragment(Attributes attrs, int tag, VR vr, JSONObject element)
```

Comment:
DataFragment isn't part of the standard, must be dcm4che specific. Also, CHC API does not return inline binary, so this code can be deleted.

Reply:
DataFragment is probably dcm4che specific, but inline binary is part of standard. But I couldn't get any from healthcare api to check how it looks there (obviously, since you just told healthcare doesn't support it).
Ok, removed.